tests/download_test.py

```diff
@@ -28,6 +29,18 @@ def test_download(self):
                     expected_return_code=0,
                 )
 
+    def test_download_no_results(self):
+        with file_utils.temp_dir(dict(malicious=self.test_eicar)) as (path, files):
+            with patch('polyswarm_api.api.PolyswarmAPI.download',
+                       return_value=[]), \
```

Comment:
If any hash values are passed into `download`, the result should never be an empty list, it should be a list of result objects. I've noticed that most of the tests mock this as the behavior in failure, but I don't think that's correct.

Reply:
`Result` objects don't exist anymore in `api`, it returns either one resource or an iterable of resources. The `PolyswarmAPI.download` yield a sequence of `LocalArtifact`. If invalid hashes were provided, one of the requests will fail and it will raise an exception. We can change this behavior to return another thing in it's place (such as an `Error` resource). When manipulating using the `api` the user does not have to worry about `http` requests, it gets abstracted away (this allow us to do implicit pagination). 

About the result being an empty list, I think it would be acceptable to return an `LocalArtifact` only for the `Artifact`s we found in our system, and ignore the rest. I am also ok with an `Error` resource in the position of the requested resource (this gets more complicated with endpoints that return multiple resources, but we can think about it). I am not ok with an `LocalArtifact` that does not hold data about an `Artifact` but holds an error.

Reply:
Thanks, I was missing that that had changed. I would strongly prefer an Error resource, as it’s imo unintuitive that an explicit command to do something returns no result sometimes (or an exception that stops all other commands from executing).

Reply:
How about if we return a mixed iterator, which could contain either `LocalArtifact` or some exception object?

Reply:
@marciogualtieri-polyswarm Yep, that is the idea of the `Error` resource. I am against returning an `Exception` directly since this is not the proper way of handling exceptions, they should be raised, not returned. I would like to have some method in the `Error` resource such as `Error.raise()` that would throw the exception that happened in the request, much like `requests.Response.raise_for_status`:

https://2.python-requests.org/en/master/api/#requests.Response.raise_for_status

This allows for the user to collect all the error methods, inspect the results and still raise the exception if he deems needed.

Reply:
Wouldn't be better to revert back to the old pattern (Ben's one) using `Result`? Then result could either contain `LocalArtifact` or some `Error`/`Exception`, which could be raised by `Result.raise()`.

Reply:
Not really, `Results` as they were held a http request information, which should be abstracted by the `api` layer. The package returns `Resources` instances and will paginate and yield as you consume the iterator. Also, it is better for users to get the resource they are asking right away rather than getting a `Result` object with a property that has the actual results they are interested into.

At some point, I would like to make the user access the resources directly, such as

```python
instance = ArtifactInstance.get(sha256='something')
for instance in ArtifactInstance.list(sha1='something with collision'):
    # do something for each one
```

This makes it very readable and easy to use, as the user knows right away what kind if instance will be returned, it is easy to inspect the code and he would get familiar with our resources.

This is much like how you would use the `Bucket` from `boto3`.

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#bucket

Reply:
As a matter of fact, we would need to provide the `api` instance to do this:

```python
instance = ArtifactInstance.get(api, sha256='something')
for instance in ArtifactInstance.list(api, sha1='something with collision'):
    # do something for each one
```

Reply:
Using the `Result` pattern, the user would always expect a `Result`, then we could use `Result.get()` (or something similar) and if there was some error (that is, there's no `LocalArtifact`, it would return an exception. That feels intuitive and readable to me.

Reply:
If you want to make the user access the resource directly, would be better to do this now? Such changes end up causing a lot of re-work later on.

Reply:
Always returning a `Result` makes it less readable. If I get some code and all that I know is that it is returning a `Result` it gives me no knowledge of what is the class that I will manipulate. Since the result will only get the instance it will hold at runtime, I can't inspect the class to see what are its methods and properties and know what to do beforehand.

```python
result = api.get_artifact(sha256='something')
instance = result.result
```

I depend on the `get_artifact()` to assign the actual `ArtifactInstance`, but I have no way of telling that it is an `ArtifactInstance` beforehand. Also, code completion tools and inpections will make it harder to navigate the source code and look at the code there.

```python
instance = ArtifactInstance.get(api, sha256='something')
```

This uses a factory pattern and the user knows it is an `ArtifactInstance` that will be returned, and likely code completion tools will be able to infer the type as well (not to mention we can add type annotation to this `get`, making tools such as `mypy` capable of validating this before runtime).

These changes can be introduced in the future without breaking or changing existing functionality in the future. People using `PolyswarmAPI` will have the same functionality, we will only add new methods that give a new way of requesting this resources directly through the resource itself.

```python
# this still works normally (although internal implementation might change to 
# use the resource pattern if we think it is necessary)
instance = api.search('sha1')

# this becomes an alternative way of getting instances with no breaking changes
instance = ArtifactInstance.list(api, sha1=''sha1'')
```

Reply:
What I know is these changes (to the models, to the abstractions, etc) always cause a lot of re-work. I'm always having to re-do work because of these changes. I would rather wait until a consensus is reached regarding design and then that we stick to it.